portfolio/src/test/java/com/google/sps/servlets/CommentServletTest.java

```diff
@@ -0,0 +1,136 @@
+ */
+ @RunWith(JUnit4.class)
+public final class CommentServletTest {
+  private final LocalServiceTestHelper helper = 
+    new LocalServiceTestHelper(new LocalDatastoreServiceTestConfig());
+
+  private static final CommentServlet commentServlet = new CommentServlet();
+  private HttpServletRequest mockRequest;
+  private HttpServletResponse mockResponse;
+  private String tokenId;
+
+  @Before
+  public void setUp() throws IOException, ServletException {
+    tokenId = getTokenId();
+    mockRequest = Mockito.mock(HttpServletRequest.class);
+    mockResponse = Mockito.mock(HttpServletResponse.class);
+    helper.setUp();
+    MockitoAnnotations.initMocks(this);
+  }
+
+  @After
+  public void tearDown() {
+    helper.tearDown();
+  }
+
+  @Test
+  public void testGet() throws IOException {
+    String data = Comment.convertToJson(generateRandomComments());
+    StringWriter stringWriter = new StringWriter();
+    stringWriter.write(data);
+    PrintWriter writer = new PrintWriter(stringWriter);
+    Mockito.when(mockResponse.getWriter()).thenReturn(writer);
+
+    commentServlet.doGet(mockRequest, mockResponse);
+    Assert.assertTrue(stringWriter.toString().contains(data));
+  }
+
+  private List<Comment> generateRandomComments() {
+    List<Comment> comments = new ArrayList<>();
+    for (int i = 1; i <= 5; i++) {
+      comments.add(new Comment(
+        Long.valueOf(i), 
+        System.currentTimeMillis(), 
+        "user" + i, 
+        "username" + i, 
+        "hello, this is a comment #" + i, 
+        0, 
+        Comment.Status.PUBLIC
+      ));
+    }
+    return comments;
+  }
+
+  @Test
+  public void testPost() throws IOException, ServletException {
+    testGet();
+    Mockito.when(mockRequest.getContentType()).thenReturn("multipart/");
+    Mockito.when(mockRequest.getParameter("id")).thenReturn(tokenId);
+
+    Collection<Part> parts = RequestDataTest.mockRequestParts();
+    Mockito.when(mockRequest.getParts()).thenReturn(parts);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Assert.assertEquals(0, datastore.prepare(new Query("Comment")).countEntities());
+    
+    commentServlet.doPost(mockRequest, mockResponse);
+
+    Assert.assertEquals(1, datastore.prepare(new Query("Comment")).countEntities());
+  }
+
+  @Test
+  public void testDelete() throws IOException, ServletException {
+    testGet();
+    testPost();
+    Mockito.when(mockRequest.getParameter("id")).thenReturn(tokenId);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Filter deletedOnlyFilter = new FilterPredicate(
+      "status", FilterOperator.EQUAL, Comment.Status.DELETED.name()
+    );
+    Query query = new Query("Comment").setFilter(deletedOnlyFilter);
+
+    Assert.assertEquals(0, datastore.prepare(query).countEntities());
+    
+    commentServlet.doDelete(mockRequest, mockResponse);
+
+    Assert.assertEquals(1, datastore.prepare(query).countEntities());
```

Comment:
Try to avoid testing behavior of your dependencies.  Represent your assumptions about their behavior with mocks.  You only want to test your own logic.

Here, for instance, you would want to mock the DatastoreService, and use an ArgumentCaptor on the write call to verify that its API is being called in the way you expect.

An alternative would be to use the Proxy design pattern to create a proxy between your servlet and the Datastore, which handles all the Datastore calls, like DatastoreWriteManagerImpl implements DatastoreWriteManager, use dependency injection to bind that implementation to your class, and then in your test bind FakeDatastoreWriteManager instead, where your fake class will have a lot of the same logic as the real thing but without actually writing to any sort of database.

Last resort would be to use a CSV file, or other small temporary local file, that you create and delete to represent the state of your DB, and initialize the DB to read/write from that instead of an actual DB.

Either way, you don't want to actually be reading/writing from a real DB in your test.